```{r knit2synapse, include=FALSE, eval=FALSE}
# library(synapseClient)
# synapseLogin()
# outFolder <- synStore(Folder(name="Integrated DE Analysis (all diff state comparisons)",
#                              parentId="syn4640410"))
# knit2synapse::knitfile2synapse("./code/Rmd/DiffStatePaper/integrate.Rmd", 
#                                owner=outFolder@properties$id, overwrite=TRUE)
```

```{r setup, include=FALSE}
library(synapseClient)
library(biomaRt) # use of biomaRt interferes w/ dplyr select ,so load it early
library(plyr)
library(dplyr)
library(data.table)
library(stringr)
library(tidyr)
library(reshape2)
library(knitr)
library(ComplexHeatmap)
library(rGithubClient)
library(igraph)
library(RColorBrewer)

set.seed(4)

outFolder <- synStore(Folder(name="Integrated DE Analysis (all diff state comparisons)",
                             parentId="syn4640410"))

thisRepo <- getRepo('kdaily/pcbc_c4_analysis', 
                    ref='branch', refName='comparisonlist')
thisScript <- getPermlink(thisRepo, 'code/Rmd/DiffStatePaper/integrate.Rmd')

knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)

synapseLogin()

# Standardized comparison names and levels
compNames <- synTableQuery("SELECT * FROM syn4483642")@values

# Diff states in a sensible, quasi time-based ordering
orderedDiffState <- c("SC", "DE", "MESO-5", "ECTO", "EB", "MESO-15", "MESO-30")

usedIds <- c()

Hs <- biomaRt::useMart("ENSEMBL_MART_ENSEMBL", 
                      host="www.ensembl.org") # use this one when biomart.org is down

Hs <- biomaRt::useDataset("hsapiens_gene_ensembl", Hs)

```

Set p-value and log fold change thresholds:

```{r setthresholds, include=TRUE, echo=TRUE}
thresh.p.mrna <- 0.0001
thresh.lfc.mrna <- 3

thresh.p.mirna <- 0.0001
thresh.lfc.mirna <- 2

thresh.p.methylation <- 0.5
thresh.lfc.methylation <- 0.4

```

## mRNA

```{r mrna}
obj <- synGet("syn4484232")
usedIds <- c(usedIds, obj@properties$id)

d.mrna <- fread(getFileLocation(obj), data.table=FALSE, 
                showProgress = FALSE) %>% 
    mutate(assay='mRNA')
```

Filter out the comparisons we're interested in at the specified significance levels.

```{r filtermrna}
d2.mrna <- d.mrna %>%
  rename(feature=GeneSymbol, comparison=Comparison) %>%
  filter(abs(logFC) > thresh.lfc.mrna,
         adj.P.value < thresh.p.mrna,
         str_detect(comparison, "^All__")) %>%
  mutate(direction=str_extract(comparison, "__.*"),
         direction=str_replace(direction, ".*__", ""),
         comparison=str_replace(comparison, "__up", ""),
         comparison=str_replace(comparison, "__down", "")) %>%
  left_join(compNames) %>%
  filter(class == "Diffname_short")

# Common set of columns of differentially expressed features
d3.mrna <- d2.mrna %>%
  dplyr::select(assay, feature) %>%
  unique

```

```{r mrna_filter2, echo=TRUE}
tmp.mrna <- d2.mrna %>% 
  dplyr::select(feature, comparison, variable1, variable2, direction) %>%
  rename(hgnc_symbol=feature) %>%
  rename(mrna_direction=direction)
```

```{r mrnatbl, echo=TRUE}
d2.mrna %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(feature)) %>%
  dcast(variable1 ~ variable2, value.var="n") %>%
  kable
```

## miRNA

```{r mirna}
obj <- synGet("syn4609631")
usedIds <- c(usedIds, obj@properties$id)

d.mirna <- fread(getFileLocation(obj), data.table=FALSE, 
                 showProgress = FALSE) %>%
  mutate(assay='miRNA')
```

Filter out the comparisons we're interested in at the specified significance levels as well.

```{r filtermirna}
d2.mirna <- d.mirna %>%
  rename(feature=GeneSymbol, comparison=Comparison) %>%
  filter(abs(logFC) > thresh.lfc.mirna,
         adj.P.value < thresh.p.mirna,
         str_detect(comparison, "^All__")) %>%
  mutate(direction=str_extract(comparison, "__.*"),
         direction=str_replace(direction, ".*__", ""),
         comparison=str_replace(comparison, "__up", ""),
         comparison=str_replace(comparison, "__down", "")) %>%
  left_join(compNames) %>%
  filter(class == "Diffname_short")
```

Table of counts of filtered differentially expressed miRNAs:

```{r mirnatbl, echo=TRUE}
d2.mirna %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(feature)) %>%
  dcast(variable1 ~ variable2, value.var="n") %>%
  kable
```

Now, use the mRNA-miRNA mapping file to add gene targets.
We only keep miRNA mapping rows where the miRNA is in our differentially expressed set.

```{r mapmirna}
# Get miRNA mapping files
DMAP.ID = 'syn3461627'
usedIds <- c(usedIds, DMAP.ID)
DMAP = fread(synGet(DMAP.ID)@filePath, data.table=F, header=F)
setnames(DMAP, c('V1','V2'), c('hsa_id','ensembl_gene_id'))

DMAP <- filter(DMAP, hsa_id %in% d2.mirna$feature)

# Get human related mapping
human_ensg2symbol = biomaRt::getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                                   filters = "ensembl_gene_id",                         
                                   values = unique(DMAP$ensembl_gene_id),
                                   mart = Hs)

DMAP <- left_join(DMAP, human_ensg2symbol, by = 'ensembl_gene_id') %>%
  rename(feature=hsa_id) %>% 
  dplyr::select(feature, hgnc_symbol)

# Add gene symbols
d3.mirna <- d2.mirna %>% left_join(DMAP) 

d4.mirna <- d3.mirna %>%
  melt(id.vars=c("assay", "feature", "comparison", "direction", "hgnc_symbol"), 
       measure.vars=c("variable1", "variable2")) %>%
  unique

# Convert to have consistent columns with other assays
d5.mirna <- d4.mirna %>%
  dplyr::select(assay, hgnc_symbol) %>%
  rename(feature=hgnc_symbol) %>%
  filter(feature != "") %>%
  unique

```

Table of counts of filtered differentially expressed miRNA-targeted genes:

```{r mirnatbl2, echo=TRUE}
d3.mirna %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(hgnc_symbol)) %>%
  dcast(variable1 ~ variable2, value.var="n") %>%
  kable
```

Now, filter miRNAs that have opposite differential expression than mRNA target and the mRNA is downregulated (miRNA up, mRNA down):

Using the miRNA-mRNA mapped data, get the same columns. 
This time, the target is the mRNA.

```{r mrna_mirna_filter2, echo=TRUE}
tmp.mirna <- d3.mirna %>% 
  dplyr::select(hgnc_symbol, feature, comparison, variable1, variable2, direction) %>%
  rename(mirna_direction=direction) %>% 
  unique
```

Combine the mrna and mirna data by gene and comparison (exact same variable comparison).
Then, filter for down-regulated (if the diff state we're interested in is listed second, then it's up-regulated)

```{r mrna_mirna_filter3, echo=TRUE}
tmp2.mirna <- tmp.mirna %>% 
  left_join(tmp.mrna) %>%
  filter(!is.na(mrna_direction), !is.na(mirna_direction))

foo.miRNA <- llply(orderedDiffState, 
                   function(x) filter(tmp2.mirna,
                                      (variable1 == x & mrna_direction == "down" & mirna_direction == "up") | (variable2 == x & mrna_direction == "up" & mirna_direction == "down")))

names(foo.miRNA) <- orderedDiffState
downreg.mrna.mirna <- ldply(foo.miRNA, I, .id="variable") %>% unique
```

### Downregulated miRNA target genes

Table of counts of differentially expressed miRNAs downregulating their targeted genes:

```{r downregmirnatbl, echo=TRUE}
downreg.mrna.mirna %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(feature)) %>%
  dcast(variable1 ~ variable2, value.var="n") %>%
  kable
```

Table of counts of the downregulated gene targets of differentially expressed miRNAs:

```{r downregmirnatbl2, echo=TRUE}
downreg.mrna.mirna %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(hgnc_symbol)) %>%
  dcast(variable1 ~ variable2, value.var="n") %>%
  kable
```

## Methylation

```{r getmethylation, echo=TRUE}
obj <- synGet("syn4527629")
d.methylation <- fread(getFileLocation(obj), data.table=FALSE, showProgress = FALSE)
usedIds <- c(usedIds, obj@properties$id)
```

Filter out the comparisons we're interested in at the specified significance levels as well.

```{r filtermethylation, echo=TRUE}
d2.methylation <- d.methylation %>% 
  mutate(assay='methylation') %>% 
  rename(feature=methProbeIDs, comparison=Comparison) %>%
  filter(abs(logFC) > thresh.lfc.methylation,
         adj.P.value < thresh.p.methylation,
         str_detect(comparison, "^All__")) %>%
  mutate(direction=str_extract(comparison, "__.*"),
         direction=str_replace(direction, ".*__", ""),
         comparison=str_replace(comparison, "__up", ""),
         comparison=str_replace(comparison, "__down", "")) %>%
  left_join(compNames) %>%
  filter(class == "Diffname_short")

d3.methylation <- d2.methylation %>%
  melt(id.vars=c("feature", "nearestTx", "assay", "direction"),
       measure.vars=c("variable1", "variable2")) %>%
  unique

# Convert to have consistent columns with other assays
d4.methylation <- d3.methylation %>%
  dplyr::select(assay, nearestTx) %>%
  rename(feature=nearestTx) %>%
  unique
```

Table of counts of filtered differentially expressed methylation probes:

```{r methyltbl, echo=TRUE}
d2.methylation %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(feature)) %>%
  dcast(variable1 ~ variable2, value.var="n") %>%
  kable
```

Table of counts of filtered differentially expressed methylation-targeted genes:

```{r methyltbl2, echo=TRUE}
d2.methylation %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(nearestTx)) %>%
  dcast(variable1 ~ variable2, value.var="n", 
        fun.aggregate=sum, margins=TRUE) %>%
  kable
```

Filter methylation probes that have opposite differential expression than mRNA target.

Using the methyl-mRNA mapped data, get the same columns. 
This time, the target is the mRNA.

```{r mrna_methyl_filter2, echo=TRUE}
tmp.methylation <- d2.methylation %>% 
  dplyr::select(nearestTx, feature, comparison, variable1, variable2, direction) %>%
  rename(hgnc_symbol=nearestTx, methylation_direction=direction) %>% 
  unique
```

Combine the mrna and mirna data by gene and comparison.

```{r mrna_methyl_filter3, echo=TRUE}
tmp.mrna.methylation <- tmp.methylation %>% 
  left_join(tmp.mrna) %>%
  filter(!is.na(mrna_direction), !is.na(methylation_direction))
```

### Down-regulated methylation target genes

Then, filter for down-regulated (if the diff state we're interested in is listed second, then it's up-regulated)

```{r mrna_methyl_filter4, echo=TRUE}

foo.methyl <- llply(orderedDiffState, 
             function(x) filter(tmp.mrna.methylation, (variable1 == x & mrna_direction == "down" & methylation_direction == "up") | (variable2 == x & mrna_direction == "up" & methylation_direction == "down")))

names(foo.methyl) <- orderedDiffState
downreg.mrna.methylation <- ldply(foo.methyl, I, .id="variable") %>%
  unique
```

Table of counts of differentially expressed miRNAs downregulating their targeted genes:

```{r downregmmethyltbl, echo=TRUE}
downreg.mrna.methylation %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(feature)) %>%
  dcast(variable1 ~ variable2, value.var="n") %>%
  kable
```

Table of counts of the downregulated gene targets of differentially expressed miRNAs:

```{r downregmethyltbl2, echo=TRUE}
downreg.mrna.methylation %>% 
  group_by(variable1, variable2) %>% 
  summarize(n=n_distinct(hgnc_symbol)) %>%
  dcast(variable1 ~ variable2, value.var="n") %>%
  kable
```


Filter based on methyl + mRNA or miRNA + mRNA differential expression.

```{r combinedfilter, echo=TRUE}
final <- rbind(d3.mrna, d5.mirna, d4.methylation) %>%
  dcast(feature ~ assay, value.var='feature', fun.aggregate=length) %>%
  filter((methylation > 0 & mRNA > 0) | (miRNA > 0 & mRNA > 0))

downreg.mrna <- d3.mrna %>% 
  filter(feature %in% c(downreg.mrna.mirna$hgnc_symbol, 
                        downreg.mrna.methylation$hgnc_symbol))

```

```{r writefinalcsv, eval=TRUE}
write.csv(final, file="integratedDEFeatures.csv")
write.csv(downreg.mrna.methylation, "downreg.mrna.methylation.csv")
write.csv(downreg.mrna.mirna, "downreg.mrna.mirna.csv")
```

```{r synstorefinalcsv, eval=TRUE}
f <- synStore(File("integratedDEFeatures.csv", parentId=outFolder@properties$id),
              used=usedIds,
              executed=thisScript)

f2 <- synStore(File("downreg.mrna.mirna.csv",
                    parentId=outFolder@properties$id),
               used=usedIds,
               executed=thisScript)

f3 <- synStore(File("downreg.mrna.methylation.csv",
                    parentId=outFolder@properties$id),
               used=usedIds,
               executed=thisScript)

# See results in `r f@properties$id`.
```


### Genes downregulated by miRNAs or methylation

Data was clustered with k-means using n = 10 cluster centers.
The data was submitted to ToppCluster for enrichment analysis.

```{r plotheatmap, fig.width=20, fig.height=15, echo=FALSE}
keepGenes <- downreg.mrna$feature

metadata <- synTableQuery("select * from syn3156503")@values %>%
  filter(Diffname_short != "") %>%
  filter(Cell_Type == "PSC") %>%  
  filter(pass_qc == "TRUE") %>%
  filter(exclude != "TRUE") %>%
  filter(C4_Karyotype_Result != "abnormal")

countObj <- synGet("syn4483934")

countData <- fread(getFileLocation((countObj)), data.table=FALSE) %>%
  rename(GeneSymbol=GeneName) %>%
  filter(GeneSymbol %in% keepGenes)

rownames(countData) <- countData$GeneSymbol

existingUIDs <- intersect(colnames(countData), metadata$UID)

metadata <- metadata %>% filter(UID %in% existingUIDs)

countData <- countData %>% dplyr::select(GeneSymbol, one_of(metadata$UID))

countData2 <- countData %>% 
  melt(id.vars="GeneSymbol", variable.name = "UID") %>%
  left_join(metadata %>% dplyr::select(UID, Diffname_short), by="UID") %>%
  group_by(GeneSymbol, Diffname_short) %>%
  summarize(expression=median(value))

countData3 <- countData2 %>% 
  dcast(GeneSymbol ~ Diffname_short) %>%
  dplyr::select(one_of(c("GeneSymbol", "SC", "DE", "MESO-5", 
                         "ECTO", "EB", "MESO-15", "MESO-30")))

rownames(countData3) <- countData3$GeneSymbol

countMat <- countData3[downreg.mrna$feature, -1]
countMat <- scale(countMat)
countMat <- t(scale(t(countMat)))

k <- 10
countMat.kmeans <- kmeans(countMat, centers=k)

dfAnnotations <- data.frame(feature=rownames(countMat)) %>% 
  mutate(cluster=factor(countMat.kmeans$cluster)) %>%
  arrange(cluster, feature)

rownames(dfAnnotations) <- dfAnnotations$feature

countMat <- countMat[as.character(dfAnnotations$feature), ]

clusterColors <- brewer.pal(k, "Set3")
names(clusterColors) <- c(1:k)

geneAnnotation <- HeatmapAnnotation(df=dfAnnotations %>% dplyr::select(cluster), which="row",
                                     col=list(cluster=clusterColors))

ht <- Heatmap(countMat, name = "log10(Expression)", 
              # col = colorRamp2(c(-2, 0, 2), c("blue", "white", "red")),
              show_row_names = FALSE, 
              show_column_names = TRUE,
              # row_names_gp = gpar(fontsize = 8),
              # column_names_gp = gpar(fontsize = 8),
              cluster_rows = FALSE,
              cluster_columns = FALSE,
              clustering_method_rows = "average",
              clustering_method_columns = "average",
              show_row_hclust = FALSE, 
              show_column_hclust = FALSE)

print(ht + geneAnnotation)
```

## Motivating Example

The results were manually inspected to identify enrichment groups specific to cluster 1, which appears to be activated in the stem cell state and is repressed over time.
I chose one particular enriched group (`Coexpression`::`M4282`::`The 'stemnes' signature: genes up-regulated and common to 6 human embryonic stem cell lines tested`), and pulled the genes related to this:
```{r echo=TRUE}
interestingGenes <- c("DDX21", "ELOVL6", "GDF3", "GJA1", "HMGA1", "HMGB2",
                      "MTHFD1", "MTHFD2", "NANOG", "PODXL", "POU5F1", "PPAT",
                      "PSIP1", "RCC2", "SEMA6A", "SEPHS1", "SET", "SLC16A1", "SNRPF")


interestingStates <- orderedDiffState # c("ECTO", "EB") # 
stateColors <- brewer.pal(length(interestingStates), 
                          name = 'Set1') # c("blue", "orange")
names(stateColors) <- interestingStates
interestingGenes
```

```{r plotgraph, fig.width=30, fig.height=30, echo=FALSE}
gd.mirna <- downreg.mrna.mirna %>%
  filter(hgnc_symbol %in% interestingGenes, 
         variable %in% interestingStates) %>% 
  dplyr::select(feature, hgnc_symbol, variable) %>% 
  unique

gd.methylation <- downreg.mrna.methylation %>%
  filter(hgnc_symbol %in% interestingGenes, 
         variable %in% interestingStates) %>% 
  dplyr::select(feature, hgnc_symbol, variable) %>% 
  unique

gd <- rbind(gd.mirna, gd.methylation)
g <- graph_from_data_frame(gd, directed=TRUE)

V(g)$type <- 1
V(g)$type[V(g)$name %in% gd.mirna[, 1]] <- 2
V(g)$type[V(g)$name %in% gd.methylation[, 1]] <- 3
E(g)$type <- as.character(gd$variable)

# toosmall <- which(graph.coreness(g) == 1 & V(g)$type == 2)
# g <- delete.vertices(g, toosmall)
# 
# toosparse <- which(degree(g, mode="in") == 1)
# g <- delete.vertices(g, toosparse)

plot(g, #layout = layout_as_bipartite,
     vertex.size=5,# vertex.label=NA,
     # edge.arrow.mode=1,
     edge.arrow.width=1,
     edge.arrow.size=0.25,
     vertex.color=c("red", "green", "blue")[V(g)$type],
     edge.color=stateColors[E(g)$type])

```
