```{r setup, include=FALSE}
library(synapseClient)
library(plyr)
library(dplyr)
library(data.table)
library(stringr)
library(tidyr)
library(reshape2)
library(ComplexHeatmap)
library(rGithubClient)
synapseLogin()

# Make sure to commit and push first!
thisRepo <- getRepo('kdaily/pcbc_c4_analysis', ref="branch", refName='diffstatepaper')
thisScript <- getPermlink(thisRepo, repositoryPath='code/Rmd/DiffStatePaper/SuppFig1.Rmd')

knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning=FALSE,
                      fig.width=20, fig.height=30)

compNames <- synTableQuery("SELECT * FROM syn4483642")@values

# Takes data and clusters cols
diffStateClusterFxn <- function(x) {
  tmp <- dcast(x, UID ~ GeneSymbol, value.var = "value")
  d <- dist(tmp, method = "euclidean")
  hclust(d, method="average")
}

# Reorders based on clustering
diffStateReorderFxn <- function(x, metadata, clusterings) {
  foo <- metadata %>% filter(Diffname_short == x)
  foo$o <-as.numeric(clusterings[[x]]$order)
  foo %>% arrange(o) %>% select(-o)  
}

# Diff states in a sensible, quasi time-based ordering
orderedDiffState <- c("SC", "EB", "DE", "MESO-5", "MESO-15", "MESO-30", "ECTO")

p.threshold.rna <- 0.0001
lfc.threshold.rna <- 2

p.threshold.mirna <- 0.0001
lfc.threshold.mirna <- 2

p.threshold.methylation <- 0.0001
lfc.threshold.methylation <- 0.5

```

```{r readdegenesdata_mrna}
obj <- synGet("syn4484232")
d <- fread(getFileLocation(obj), data.table=FALSE)

# Filter out the comparisons we're interested in
# And significance levels as well
d2 <- d %>% 
  rename(comparison=Comparison) %>%
  filter(abs(logFC) > lfc.threshold.rna,
         adj.P.value < p.threshold.rna,
         str_detect(comparison, "^All__")) %>%
  mutate(direction=str_extract(comparison, "__.*"),
         direction=str_replace(direction, ".*__", ""),
         comparison=str_replace(comparison, "__up", ""),
         comparison=str_replace(comparison, "__down", "")) %>%
  left_join(compNames) %>%
  filter(class == "Diffname_short")
```

```{r readmetadata_mrna}
metadata <- synTableQuery("select * from syn3156503")@values %>%
  filter(Diffname_short != "") %>%
  filter(Cell_Type == "PSC") %>%  
  filter(pass_qc == "TRUE") %>%
  filter(exclude != "TRUE") %>%
  filter(C4_Karyotype_Result != "abnormal")
```


```{r getcountdata_mrna}
countObj <- synGet("syn4483934")

countData <- fread(getFileLocation((countObj)), data.table=FALSE) %>%
  rename(GeneSymbol=GeneName) %>%
  filter(GeneSymbol %in% d2$GeneSymbol)

rownames(countData) <- countData$GeneSymbol

existingUIDs <- intersect(colnames(countData), metadata$UID)

metadata <- metadata %>% filter(UID %in% existingUIDs)

countData <- countData %>% select(GeneSymbol, one_of(metadata$UID))

countData2 <- countData
rownames(countData2) <- countData2$GeneSymbol

# assayData <- as.matrix(countData[, -1])
# phenoData <- metadata
# rownames(phenoData) <- metadata$UID
# 
# eset <- ExpressionSet(assayData=assayData,
#                       phenoData=AnnotatedDataFrame(phenoData),
#                       featureData=AnnotatedDataFrame(countData[, c("GeneSymbol"), drop=F]))

```

```{r cluster_mRNA}
countData3 <- melt(countData2, id.vars = "GeneSymbol") %>%
  rename(UID=variable) %>% left_join(metadata[, c("UID", "Diffname_short")])

clusterings <- dlply(countData3, .(Diffname_short), diffStateClusterFxn)

metadataReordered <- ldply(orderedDiffState, 
                           diffStateReorderFxn, 
                           metadata=metadata, clusterings=clusterings)

countMat.rna <- countData2[, -1]
countMat.rna <- scale(countMat.rna)
countMat.rna <- t(scale(t(countMat.rna)))

countMat.rna <- countMat.rna[, metadataReordered$UID]

```

```{r plot_mrna}
dfAnnotations <- metadataReordered %>% 
  rename(DiffState=Diffname_short) %>%
  select(DiffState)

samplesAnnotation <- HeatmapAnnotation(df=dfAnnotations,
                                       col=list(DiffState=c("EB"="red", "ECTO"="blue", 
                                                            "SC"="orange", "DE"="yellow",
                                                            "MESO-5"="black", 
                                                            "MESO-15"="green",
                                                            "MESO-30"="pink")))

pdf('SuppFig1_mRNA.pdf')
ht <- Heatmap(countMat.rna, name = "log10(Expression)", 
              # col = colorRamp2(c(-2, 0, 2), c("blue", "white", "red")),
              show_row_names = FALSE, 
              show_column_names = FALSE,
              # row_names_gp = gpar(fontsize = 8),
              # column_names_gp = gpar(fontsize = 8),
              cluster_rows = TRUE,
              clustering_method_rows = "average",
              show_row_hclust = FALSE, 
              show_column_hclust = FALSE, 
              top_annotation = samplesAnnotation)

print(ht)
dev.off()

suppfig1.mRNA <- synStore(File('SuppFig1_mRNA.pdf', parentId='syn4904956'),
                          activityName="Create figure",
                          used=c('syn4483934', 'syn4484232', 'syn3156503', 'syn4483642'), 
                          executed=thisScript)

print(ht)

```

```{r readdegenesdata_mirna}
obj <- synGet("syn4609631")
d <- fread(getFileLocation(obj), data.table=FALSE)

# Filter out the comparisons we're interested in
# And significance levels as well
d2 <- d %>% 
  rename(comparison=Comparison) %>%
  filter(abs(logFC) > lfc.threshold.mirna,
         adj.P.value < p.threshold.mirna,
         str_detect(comparison, "^All__")) %>%
  mutate(direction=str_extract(comparison, "__.*"),
         direction=str_replace(direction, ".*__", ""),
         comparison=str_replace(comparison, "__up", ""),
         comparison=str_replace(comparison, "__down", "")) %>%
  left_join(compNames) %>%
  filter(class == "Diffname_short")
```

```{r readmetadata_mirna}
metadata <- synTableQuery("select * from syn3219876")@values %>%
  filter(Diffname_short != "") %>%
  filter(Cell_Type == "PSC") %>%  
  filter(pass_qc == "TRUE") %>%
  filter(exclude != "TRUE") %>%
  filter(C4_Karyotype_Result != "abnormal")
```

```{r getcountdata_mirna}
countObj <- synGet("syn3355993")

countData <- fread(getFileLocation((countObj)), data.table=FALSE) %>%
  rename(GeneSymbol=id) %>%
  filter(GeneSymbol %in% d2$GeneSymbol)

existingUIDs <- intersect(colnames(countData), metadata$UID)

metadata <- metadata %>% filter(UID %in% existingUIDs)

countData <- countData %>%
  dplyr::select(one_of(c("GeneSymbol", metadata$UID)))

countData2 <- countData
rownames(countData2) <- countData2$GeneSymbol

countData2[, -1] <- log10(countData2[, -1] + 0.001)

```

```{r cluster_mirna}
countData3 <- melt(countData2, id.vars = "GeneSymbol") %>%
  rename(UID=variable) %>% left_join(metadata[, c("UID", "Diffname_short")])

clusterings <- dlply(countData3, .(Diffname_short), diffStateClusterFxn)

metadataReordered <- ldply(orderedDiffState, 
                           diffStateReorderFxn, 
                           metadata=metadata, clusterings=clusterings)

countMat.miRNA <- countData2[, -1]
countMat.miRNA <- scale(countMat.miRNA)
countMat.miRNA <- t(scale(t(countMat.miRNA)))

countMat.miRNA <- countMat.miRNA[, metadataReordered$UID]

```

```{r plot_mirna}
dfAnnotations <- metadataReordered %>% 
  rename(DiffState=Diffname_short) %>%
  select(DiffState)

samplesAnnotation <- HeatmapAnnotation(df=dfAnnotations,
                                       col=list(DiffState=c("EB"="red", "ECTO"="blue", 
                                                            "SC"="orange", "DE"="yellow",
                                                            "MESO-5"="black", 
                                                            "MESO-15"="green",
                                                            "MESO-30"="pink")))

pdf('SuppFig1_miRNA.pdf')
ht <- Heatmap(countMat.miRNA, name = "log10(Expression)", 
              show_row_names = FALSE, 
              show_column_names = FALSE,
              cluster_rows = TRUE,
              clustering_method_rows = "average",
              show_row_hclust = FALSE, 
              show_column_hclust = FALSE, 
              top_annotation = samplesAnnotation)

print(ht)
dev.off()

suppfig1.miRNA <- synStore(File('SuppFig1_miRNA.pdf', parentId='syn4904956'),
                           activityName="Create figure",
                           used=c('syn3355993', 'syn3219876', 'syn4609631', 'syn4483642'), 
                           executed=thisScript)

print(ht)

```

```{r readdegenesdata_methylation}
obj <- synGet("syn4527629")
d <- fread(getFileLocation(obj), data.table=FALSE)

# Filter out the comparisons we're interested in
# And significance levels as well
d2 <- d %>% 
  rename(GeneSymbol=methProbeIDs, comparison=Comparison) %>%
  filter(abs(logFC) > lfc.threshold.methylation,
         adj.P.value < p.threshold.methylation,
         str_detect(comparison, "^All__")) %>%
  mutate(direction=str_extract(comparison, "__.*"),
         direction=str_replace(direction, ".*__", ""),
         comparison=str_replace(comparison, "__up", ""),
         comparison=str_replace(comparison, "__down", "")) %>%
  left_join(compNames) %>%
  filter(class == "Diffname_short")
```

```{r readmetadata_methylation}
metadata <- synTableQuery("select * from syn3156828")@values %>%
  filter(Diffname_short != "") %>%
  filter(Cell_Type == "PSC") %>%  
  filter(C4_Karyotype_Result != "abnormal")
```

```{r getbetadata_methylation}
countObj <- synGet("syn4487642")

countData <- fread(getFileLocation((countObj)), data.table=FALSE) %>%
  rename(GeneSymbol=methProbeID) %>%
  filter(GeneSymbol %in% d2$GeneSymbol)

existingUIDs <- intersect(colnames(countData), metadata$UID)

metadata <- metadata %>% filter(UID %in% existingUIDs)

countData <- countData %>%
  dplyr::select(one_of(c("GeneSymbol", metadata$UID)))

countData2 <- countData
rownames(countData2) <- countData2$GeneSymbol

# countData2[, -1] <- log10(countData2[, -1] + 0.001)

```

```{r cluster_methylation}
countData3 <- melt(countData2, id.vars = "GeneSymbol") %>%
  rename(UID=variable) %>% left_join(metadata[, c("UID", "Diffname_short")])

clusterings <- dlply(countData3, .(Diffname_short), diffStateClusterFxn)

metadataReordered <- ldply(orderedDiffState, 
                           diffStateReorderFxn, 
                           metadata=metadata, clusterings=clusterings)

countMat.methylation <- countData2[, -1]
countMat.methylation <- scale(countMat.methylation)
countMat.methylation <- t(scale(t(countMat.methylation)))

countMat.methylation <- countMat.methylation[, metadataReordered$UID]

```

```{r plot_methylation}
dfAnnotations <- metadataReordered %>% 
  rename(DiffState=Diffname_short) %>%
  select(DiffState)

samplesAnnotation <- HeatmapAnnotation(df=dfAnnotations,
                                       col=list(DiffState=c("EB"="red", "ECTO"="blue", 
                                                            "SC"="orange", "DE"="yellow",
                                                            "MESO-5"="black", 
                                                            "MESO-15"="green",
                                                            "MESO-30"="pink")))

pdf('SuppFig1_methylation.pdf')
ht <- Heatmap(countMat.methylation, name = "log10(Expression)", 
              # col = colorRamp2(c(-2, 0, 2), c("blue", "white", "red")),
              show_row_names = FALSE, 
              show_column_names = FALSE,
              # row_names_gp = gpar(fontsize = 8),
              # column_names_gp = gpar(fontsize = 8),
              cluster_rows = TRUE,
              cluster_columns = FALSE,
              clustering_method_rows = "average",
              clustering_method_columns = "average",
              show_row_hclust = FALSE, 
              show_column_hclust = FALSE, 
              top_annotation = samplesAnnotation)

print(ht)
dev.off()

suppfig1.miRNA <- synStore(File('SuppFig1_methylation.pdf', parentId='syn4904956'),
                           activityName="Create figure",
                           used=c('syn3355993', 'syn3219876', 'syn4609631', 'syn4483642'), 
                           executed=thisScript)

```

## Supplementary Figure 1a

The log CPM (counts per million) matrix for RNA-seq (syn4483934), $log_10$ count matrix for miRNA-seq (syn3355993) and $\beta$ values for methylation (syn4527629) were used for visualization of global patterns across differentiation states. Only differentially expressed genes between any pair of differentiation states were used (mRNA: syn4484232; miRNA:; methylation: ; mRNA log fold change $>`r lfc.threshold.rna`$ and adjusted $p$-value $<`r p.threshold.rna`$; $n=`r nrow(countMat)`$ genes; miRNA log fold change $>`r lfc.threshold.mirna`$ and adjusted $p$-value $<`r p.threshold.mirna`$; $n=`r nrow(countMat.miRNA)`$ miRNAs; methylation log fold change $>`r lfc.threshold.methylation`$ and adjusted $p$-value $<`r p.threshold.methylation`$; $n=`r nrow(countMat.methylation)`$ probes;). For each assay, features (e.g., genes, miRNA's and methylation probes, respectively) were clustered using hierarchical clustering (Euclidean distance and average agglomeration method). The samples were clustered using hierarchical clustering (Euclidean distance and average agglomeration method) within each differentiation state. The differentiation states were then ordered temporally with respect to the level of differentiation. 
